src/Type/UnionType.php

```diff
@@ -206,6 +206,8 @@ public function describe(VerbosityLevel $level): string
 			foreach ($types as $type) {
 				if ($type instanceof ClosureType || $type instanceof CallableType || $type instanceof TemplateUnionType) {
 					$typeNames[] = sprintf('(%s)', $type->describe($level));
+				} elseif ($type instanceof TemplateType && ($level->isTypeOnly() || $level->isValue())) {
```

Comment:
I was doubting to do this regardless of level, but for precise/cache the template scope and variance is appended which already makes it unambiguous.